articles/execute_process_refactor.md

```diff
@@ -0,0 +1,596 @@
+
+#### Methods
+
+| Name    | Description                                                  |
+| ------- | ------------------------------------------------------------ |
+| execute | Override. Adds subscription event handlers to listen for ROS node lifecycle state changes; this handler will trigger appropriate events from this class when transitions are detected. |
+
+#### Events
+
+##### Handled Events
+
+In addition to events handled by the `ExecuteProcess` superclass:
+
+| Event Type                            | Description                                                  |
+| ------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.lifecycle.ChangeState` | This event can be targeted to a single lifecycle node, or more than one, or even all lifecycle nodes, and it requests the targeted nodes to change state, see its documentation for more details |
+
+##### Emitted Events
+
+In addition to events emitted by the `ExecuteProcess` superclass:
+
+| Event Type                                | Description                                                  |
+| ----------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.lifecycle.StateTransition` | This event is emitted when a message is published to the "`/<node_name>/transition_event`" topic, indicating the lifecycle node represented by this action changed state |
+
+## Considerations for Refactored Classes
+
+The primary goal of the refactor is to attempt to split the definition aspect of processes/nodes apart from the logic which executes those elements.
+`ComposableNode` provides a basis for similar breakouts of other elements.
+
+In general, existing classes should have similar analogs in the refactor to ensure that existing classes can be replaced with backwards-compatible replacements that use the new structure.
+These replacements can be implemented to warn the user that the older definitions are deprecated, and suggest moving to the newer refactored style.
+
+Further consideration will be required for alternate frontends, such as XML, to determine how to implement these changes with streamlined syntax where possible.
+
+One difficulty when looking forward to allowing functionality such as remote and containerized execution is that the target environment is likely different than the local environment.
+Different packages may be installed, different environment variables may be defined, and so forth.
+An subclass implementation of the proposed `launch.descriptions.Executable` class may choose to override the base `apply_context` method to implement additional layer(s) of substitutions appropriate to its destination contexts, such as remote installation paths or environment variables.
```

Comment:
I'm still unsure if the concept of overriding `apply_context` is needed.
I don't see much benefit on it

Reply:
The intent of allowing overrides of `apply_context` is so an executable in a more specific domain can perform domain-specific substitutions. Basically, at this level, we're in a context that has no knowledge of ROS or any ROS-specific concepts. We can do some level of substitution here, but can't take everything into consideration. To support those ROS-specific concepts, we have the `launch_ros.descriptions.RosExecutable` class. It would override `apply_context` so that it could pass substitution information down to the `launch_ros.descriptions.Node` objects, and possibly perform other actions necessary for preparing the ROS executable to run.

My fear is that if we don't do things this way, then we either have to make the `launch` namespace aware of ROS-specific concepts like nodes, or we alternately need to come up with some other more generic framework to ensure all aspects of the definition have an opportunity to do pre-launch setup (primarily substitutions). If there's another approach I'm missing that doesn't require the override, I'd be very happy to discuss it and look for a different take.

Reply:
>  The intent of allowing overrides of apply_context is so an executable in a more specific domain can perform domain-specific substitutions

Can you give an example of a domain specific substitution?
I'm trying to understand the use case, but I'm failing to see how this will help.

> so that it could pass substitution information down to the launch_ros.descriptions.Node objects, and possibly perform other actions necessary for preparing the ROS executable to run.

Can you give an example of what "other actions" can be.

---

You can still go ahead with a proposed implementation with this comment unsolved, it's not a big deal.
But I want to understand how this detail is supposed to help.

Reply:
@ivanpauno It might be best to show you the current functionality I'm trying to find a place for. In the current setup, `launch_ros.actions.Node` is a subclass of `launch.actions.ExecuteProcess`. `ExecuteProcess` defines the `execute` method, which is overridden by `Node`. Part of what `Node` does with the override is applying the various substitutions by calling `_perform_substitutions(LaunchContext)`. Some of those substitutions are ROS-specific, some may not be. (Honestly, I'm going to have to be very careful in that area to make sure I split things out appropriately.) Those substitutions still need to be possible after the refactor.

The purpose of *this* call is to allow a subclass (in this case, `launch_ros.descriptions.RosExecutable`) to inject any additional substitution logic. For instance, the proposed `launch.descriptions.Executable` has absolutely no reason to know anything at all about a `launch_ros.descriptions.Node`'s Name property, but it still needs to have an opportunity to perform substitutions. I'll admit to not being 100% on all the functionality going on down there, but my feeling is that I need some path to make sure everything can be covered. This was my best effort at ensuring that path exists when it's needed.

Reply:
Oh, I see what you mean now.

You need a way to "normalize" all `launch.descriptions.Executable` derived classes, in a way that `launch.actions.ExecuteLocal` or other actions can actually execute them.

The proposal makes sense to me!!